bug: MLX models fail to load in 0.7.9 on macOS Apple Silicon (Failed to load the default metallib), but work in 0.7.8

**Version:** 0.7.9

## Describe the Bug

MLX models fail to load on macOS Tahoe Apple Silicon (Mac mini M4) in Jan 0.7.9.

On my machine, the same MLX model worked in Jan 0.7.8, but after updating to 0.7.9 it consistently crashes during model startup with:

MLX error: Failed to load the default metallib. library not found

Jan shows:
{"code":"MLX_PROCESS_ERROR","details":"","message":"The MLX model process encountered an unexpected error."}

This seems to be a regression in 0.7.9’s MLX runtime/packaging, not a problem with the model itself.

## Steps to Reproduce
1. Install Jan 0.7.9 on macOS Tahoe Apple Silicon.
2. Load an MLX model such as gemma-3n-E4B-it-lm-4bit.
3. Try to start a chat / load the model.
4. The model fails immediately with MLX_PROCESS_ERROR.

## Screenshots / Logs

[2026-03-23][14:28:07][tauri_plugin_mlx::commands][INFO] Attempting to launch MLX server at path: "/Applications/Jan.app/Contents/Resources/resources/bin/mlx-server"
[2026-03-23][14:28:07][tauri_plugin_mlx::commands][INFO] Using MLX configuration: MlxConfig { ctx_size: 8192 }
[2026-03-23][14:28:07][tauri_plugin_mlx::commands][INFO] MLX server arguments: ["--model", "/Users/dubis/Library/Application Support/Jan/data/mlx/models/gemma-3n-E4B-it-lm-4bit/model.safetensors", "--port", "3858", "--model-id", "gemma-3n-E4B-it-lm-4bit", "--ctx-size", "8192", "--api-key", "..."]
[2026-03-23][14:28:08][tauri_plugin_mlx::commands][INFO] [mlx stdout] MLX error: Failed to load the default metallib. library not found library not found library not found library not found at /Users/runner/work/jan/jan/mlx-server/.build/checkouts/mlx-swift/Source/Cmlx/mlx-c/mlx/c/memory.cpp:69
[2026-03-23][14:28:08][tauri_plugin_mlx::commands][ERROR] MLX server exited with error code ExitStatus(unix_wait_status(65280))
[2026-03-23][14:28:08][webview:error@asset://localhost/...][ERROR] Error loading MLX model: {"code":"MLX_PROCESS_ERROR","details":"","message":"The MLX model process encountered an unexpected error."}

Additional info:
- mlx-server exists at:
/Applications/Jan.app/Contents/Resources/resources/bin/mlx-server
- After a fresh reinstall of 0.7.9, running:
find "/Applications/Jan.app" -iname "*.metallib"
returns nothing
- Wiping ~/Library/Application Support/Jan and reinstalling did not fix it
- The same model worked (and still does, so I downgraded) in Jan 0.7.8

## Operating System
- macOS Tahoe

--- a/Makefile
+++ b/Makefile
@@ -149,11 +149,18 @@ ifeq ($(DETECTED_OS),Darwin)
 	mkdir -p src-tauri/resources/bin; \
 	echo "Copying mlx-server from $$BUILD_DIR..."; \
 	cp "$$BUILD_DIR/mlx-server" src-tauri/resources/bin/mlx-server; \
-	if [ -d "$$BUILD_DIR/mlx-swift_Cmlx.bundle" ]; then \
-		cp -r "$$BUILD_DIR/mlx-swift_Cmlx.bundle" src-tauri/resources/bin/; \
-	else \
-		mkdir -p src-tauri/resources/bin/mlx-swift_Cmlx.bundle; \
+	CMLX_BUNDLE=$$(find "$$BUILD_DIR" -maxdepth 4 -type d -name '*Cmlx*.bundle' -print -quit); \
+	if [ -z "$$CMLX_BUNDLE" ]; then \
+		CMLX_BUNDLE=$$(find "$$BUILD_DIR" -maxdepth 5 -type f -name 'default.metallib' -print -quit | xargs -I{} dirname {}); \
 	fi; \
+	if [ -z "$$CMLX_BUNDLE" ] || [ ! -f "$$CMLX_BUNDLE/default.metallib" ]; then \
+		echo "Error: mlx-swift Cmlx resource bundle with default.metallib not found under $$BUILD_DIR"; \
+		find "$$BUILD_DIR" -maxdepth 5 -name '*.bundle' -o -name 'default.metallib' 2>/dev/null; \
+		exit 1; \
+	fi; \
+	echo "Copying Cmlx bundle from $$CMLX_BUNDLE..."; \
+	rm -rf src-tauri/resources/bin/mlx-swift_Cmlx.bundle; \
+	cp -r "$$CMLX_BUNDLE" src-tauri/resources/bin/mlx-swift_Cmlx.bundle; \
 	chmod +x src-tauri/resources/bin/mlx-server; \
 	echo "MLX server built and copied successfully"; \
 	echo "Checking for code signing identity..."; \
@@ -161,10 +168,12 @@ ifeq ($(DETECTED_OS),Darwin)
 	if [ -n "$$SIGNING_IDENTITY" ]; then \
 		echo "Signing mlx-server with identity: $$SIGNING_IDENTITY"; \
 		codesign --force --options runtime --timestamp --sign "$$SIGNING_IDENTITY" src-tauri/resources/bin/mlx-server; \
-		if [ -d "src-tauri/resources/bin/mlx-swift_Cmlx.bundle" ]; then \
-			echo "Signing mlx-swift_Cmlx.bundle..."; \
-			codesign --force --options runtime --timestamp --sign "$$SIGNING_IDENTITY" --deep src-tauri/resources/bin/mlx-swift_Cmlx.bundle; \
+		if [ ! -f "src-tauri/resources/bin/mlx-swift_Cmlx.bundle/default.metallib" ]; then \
+			echo "Error: mlx-swift_Cmlx.bundle is missing default.metallib; refusing to sign an empty bundle"; \
+			exit 1; \
 		fi; \
+		echo "Signing mlx-swift_Cmlx.bundle..."; \
+		codesign --force --options runtime --timestamp --sign "$$SIGNING_IDENTITY" --deep src-tauri/resources/bin/mlx-swift_Cmlx.bundle; \
 		echo "Code signing completed successfully"; \
 	else \
 		echo "Warning: No Developer ID Application identity found. Skipping code signing (notarization will fail)."; \

--- a/Makefile
+++ b/Makefile
@@ -149,15 +149,19 @@ ifeq ($(DETECTED_OS),Darwin)
 	mkdir -p src-tauri/resources/bin; \
 	echo "Copying mlx-server from $$BUILD_DIR..."; \
 	cp "$$BUILD_DIR/mlx-server" src-tauri/resources/bin/mlx-server; \
-	CMLX_BUNDLE=$$(find "$$BUILD_DIR" -maxdepth 4 -type d -name '*Cmlx*.bundle' -print -quit); \
-	if [ -z "$$CMLX_BUNDLE" ]; then \
-		CMLX_BUNDLE=$$(find "$$BUILD_DIR" -maxdepth 5 -type f -name 'default.metallib' -print -quit | xargs -I{} dirname {}); \
-	fi; \
-	if [ -z "$$CMLX_BUNDLE" ] || [ ! -f "$$CMLX_BUNDLE/default.metallib" ]; then \
-		echo "Error: mlx-swift Cmlx resource bundle with default.metallib not found under $$BUILD_DIR"; \
-		find "$$BUILD_DIR" -maxdepth 5 -name '*.bundle' -o -name 'default.metallib' 2>/dev/null; \
+	BUILD_ROOT="$$(cd mlx-server && pwd)/.build"; \
+	METALLIB=$$(find "$$BUILD_ROOT" -type f -name 'default.metallib' -print 2>/dev/null | head -1); \
+	if [ -z "$$METALLIB" ]; then \
+		echo "Error: default.metallib not found anywhere under $$BUILD_ROOT"; \
+		echo "--- *.bundle under $$BUILD_ROOT ---"; \
+		find "$$BUILD_ROOT" -name '*.bundle' 2>/dev/null; \
+		echo "--- *.metallib under $$BUILD_ROOT ---"; \
+		find "$$BUILD_ROOT" -name '*.metallib' 2>/dev/null; \
+		echo "--- top-level .build layout ---"; \
+		ls -la "$$BUILD_ROOT" 2>/dev/null; \
 		exit 1; \
 	fi; \
+	CMLX_BUNDLE=$$(dirname "$$METALLIB"); \
 	echo "Copying Cmlx bundle from $$CMLX_BUNDLE..."; \
 	rm -rf src-tauri/resources/bin/mlx-swift_Cmlx.bundle; \
 	cp -r "$$CMLX_BUNDLE" src-tauri/resources/bin/mlx-swift_Cmlx.bundle; \

--- a/Makefile
+++ b/Makefile
@@ -139,32 +139,28 @@ endif
 build-mlx-server:
 ifeq ($(DETECTED_OS),Darwin)
 	@echo "Building MLX server for Apple Silicon..."
-	cd mlx-server && swift build -c release
-	@echo "Copying build products..."
-	@BUILD_DIR=$$(cd mlx-server && swift build -c release --show-bin-path); \
-	if [ -z "$$BUILD_DIR" ]; then \
-		echo "Error: Could not find build products"; \
+	# mlx-swift's Metal shaders are compiled by the PrepareMetalShaders
+	# plugin, which only runs under Xcode -- `swift build` produces a
+	# binary with no default.metallib and the app fails at runtime. See
+	# https://github.com/ml-explore/mlx-swift README ("SwiftPM (command
+	# line) cannot build the Metal shaders").
+	cd mlx-server && xcodebuild build -scheme mlx-server -destination 'platform=OS X' -configuration Release OTHER_LDFLAGS="-dead_strip"
+	@echo "Finding build products..."
+	@DERIVED_DATA=$$(find ~/Library/Developer/Xcode/DerivedData/mlx-server-*/Build/Products/Release -maxdepth 0 2>/dev/null | head -1); \
+	if [ -z "$$DERIVED_DATA" ] || [ ! -f "$$DERIVED_DATA/mlx-server" ]; then \
+		echo "Error: Could not find xcodebuild products under DerivedData"; \
+		exit 1; \
+	fi; \
+	if [ ! -f "$$DERIVED_DATA/mlx-swift_Cmlx.bundle/default.metallib" ]; then \
+		echo "Error: $$DERIVED_DATA/mlx-swift_Cmlx.bundle/default.metallib missing -- PrepareMetalShaders did not run"; \
+		ls -la "$$DERIVED_DATA" 2>/dev/null; \
 		exit 1; \
 	fi; \
 	mkdir -p src-tauri/resources/bin; \
-	echo "Copying mlx-server from $$BUILD_DIR..."; \
-	cp "$$BUILD_DIR/mlx-server" src-tauri/resources/bin/mlx-server; \
-	BUILD_ROOT="$$(cd mlx-server && pwd)/.build"; \
-	METALLIB=$$(find "$$BUILD_ROOT" -type f -name 'default.metallib' -print 2>/dev/null | head -1); \
-	if [ -z "$$METALLIB" ]; then \
-		echo "Error: default.metallib not found anywhere under $$BUILD_ROOT"; \
-		echo "--- *.bundle under $$BUILD_ROOT ---"; \
-		find "$$BUILD_ROOT" -name '*.bundle' 2>/dev/null; \
-		echo "--- *.metallib under $$BUILD_ROOT ---"; \
-		find "$$BUILD_ROOT" -name '*.metallib' 2>/dev/null; \
-		echo "--- top-level .build layout ---"; \
-		ls -la "$$BUILD_ROOT" 2>/dev/null; \
-		exit 1; \
-	fi; \
-	CMLX_BUNDLE=$$(dirname "$$METALLIB"); \
-	echo "Copying Cmlx bundle from $$CMLX_BUNDLE..."; \
+	echo "Copying mlx-server from $$DERIVED_DATA..."; \
+	cp "$$DERIVED_DATA/mlx-server" src-tauri/resources/bin/mlx-server; \
 	rm -rf src-tauri/resources/bin/mlx-swift_Cmlx.bundle; \
-	cp -r "$$CMLX_BUNDLE" src-tauri/resources/bin/mlx-swift_Cmlx.bundle; \
+	cp -r "$$DERIVED_DATA/mlx-swift_Cmlx.bundle" src-tauri/resources/bin/; \
 	chmod +x src-tauri/resources/bin/mlx-server; \
 	echo "MLX server built and copied successfully"; \
 	echo "Checking for code signing identity..."; \

--- a/Makefile
+++ b/Makefile
@@ -151,9 +151,10 @@ ifeq ($(DETECTED_OS),Darwin)
 		echo "Error: Could not find xcodebuild products under DerivedData"; \
 		exit 1; \
 	fi; \
-	if [ ! -f "$$DERIVED_DATA/mlx-swift_Cmlx.bundle/default.metallib" ]; then \
-		echo "Error: $$DERIVED_DATA/mlx-swift_Cmlx.bundle/default.metallib missing -- PrepareMetalShaders did not run"; \
-		ls -la "$$DERIVED_DATA" 2>/dev/null; \
+	METALLIB=$$(find "$$DERIVED_DATA/mlx-swift_Cmlx.bundle" -name 'default.metallib' -print -quit 2>/dev/null); \
+	if [ -z "$$METALLIB" ]; then \
+		echo "Error: default.metallib missing under $$DERIVED_DATA/mlx-swift_Cmlx.bundle -- PrepareMetalShaders did not run"; \
+		find "$$DERIVED_DATA/mlx-swift_Cmlx.bundle" -maxdepth 4 2>/dev/null; \
 		exit 1; \
 	fi; \
 	mkdir -p src-tauri/resources/bin; \
@@ -168,8 +169,8 @@ ifeq ($(DETECTED_OS),Darwin)
 	if [ -n "$$SIGNING_IDENTITY" ]; then \
 		echo "Signing mlx-server with identity: $$SIGNING_IDENTITY"; \
 		codesign --force --options runtime --timestamp --sign "$$SIGNING_IDENTITY" src-tauri/resources/bin/mlx-server; \
-		if [ ! -f "src-tauri/resources/bin/mlx-swift_Cmlx.bundle/default.metallib" ]; then \
-			echo "Error: mlx-swift_Cmlx.bundle is missing default.metallib; refusing to sign an empty bundle"; \
+		if ! find src-tauri/resources/bin/mlx-swift_Cmlx.bundle -name 'default.metallib' -print -quit 2>/dev/null | grep -q .; then \
+			echo "Error: staged mlx-swift_Cmlx.bundle is missing default.metallib; refusing to sign an empty bundle"; \
 			exit 1; \
 		fi; \
 		echo "Signing mlx-swift_Cmlx.bundle..."; \

--- a/web-app/src/hooks/useModelProvider.ts
+++ b/web-app/src/hooks/useModelProvider.ts
@@ -38,7 +38,13 @@ export const useModelProvider = create<ModelProviderState>()(
       },
       setProviders: (providers) =>
         set((state) => {
+          // MLX is Apple-Silicon only; drop it on every other platform so it
+          // can't be reactivated, queried, or shown anywhere in the UI.
+          providers = IS_MACOS
+            ? providers
+            : providers.filter((e) => e.provider !== 'mlx')
           const existingProviders = state.providers
+            .filter((e) => IS_MACOS || e.provider !== 'mlx')
             // Filter out legacy llama.cpp provider for migration
             // Can remove after a couple of releases
             .filter((e) => e.provider !== 'llama.cpp')

